h/util/session_tracker.py

```diff
@@ -0,0 +1,84 @@
+from copy import copy
+from enum import Enum
+
+from sqlalchemy.event import listen
+from sqlalchemy.inspection import inspect
+
+
+class ObjectState(Enum):
+    ADDED = 'added'
+    DELETED = 'deleted'
+    CHANGED = 'changed'
+
+
+def db_object_key(db_obj):
+    """
+    Return a serializable identifier for a mapped DB object.
+
+    This is intended to be useful in eg. log output.
+    """
+    primary_key = inspect(db_obj).identity
+    return (type(db_obj), primary_key)
+
+
+class Tracker(object):
+    """
+    Observer which tracks whether a SQLAlchemy `Session` has uncommitted changes.
+    """
+
+    def __init__(self, session):
+        self.session = session
+
+        # Changes which have been flushed to the database (and thus not present
+        # in `session.new`, `session.dirty` or `session.deleted) but have not
+        # yet been _committed_.
+        self._flushed_changes = {}
+
+        listen(session, 'after_flush', self._after_flush)
+        listen(session, 'after_commit', self._after_commit)
+        listen(session, 'after_rollback', self._after_rollback)
+
+    def uncommitted_changes(self):
+        """
+        Return a list of changes to the `Session` which have not yet been
+        _committed_ to the DB.
+
+        The result is a list of (model class, primary key, ObjectState) tuples.
+        """
+        changed = copy(self._flushed_changes)
+        changed.update(self._unflushed_changes())
+
+        return [(model_cls, pk, state) for ((model_cls, pk), state) in changed.iteritems()]
+
+    def _unflushed_changes(self):
+        """
+        Return a map of changes which have not yet been flushed to the DB.
+
+        In the context of the "after_flush" event handler, this returns changes
+        which have just been flushed.
+
+        If an object goes through multiple states in the same session (eg.
+        added, then flushed, then changed) then only the last state for a given
+        object is recorded.
+        """
+        changes = {}
+
+        for obj in self.session.new:
+            changes[db_object_key(obj)] = ObjectState.ADDED
+
+        for obj in self.session.dirty:
+            changes[db_object_key(obj)] = ObjectState.CHANGED
+
+        for obj in self.session.deleted:
+            changes[db_object_key(obj)] = ObjectState.DELETED
+
+        return changes
+
+    def _after_rollback(self, session):
```

Comment:
Might it be worth underscoring out the `session` arguments here and in the method below, to make it clearer that they’re not used in the methods themselves? You could even go a step further and `*args` them, to make it even more apparent that you don’t care what’s being passed into these handlers.